auth/oauth2.go

```diff
@@ -0,0 +1,116 @@
+// Copyright 2023 Contributors to the Veraison project.
+// SPDX-License-Identifier: Apache-2.0
+package auth
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"net/url"
+	"time"
+
+	"github.com/mitchellh/mapstructure"
+	"golang.org/x/oauth2"
+)
+
+type Oauth2Authenticator struct {
+	TokenURL     string
+	ClientID     string
+	ClientSecret string
+	Username     string
+	Password     string
+
+	Token *oauth2.Token
+}
+
+func (o *Oauth2Authenticator) Configure(cfg map[string]interface{}) error {
+	decoded := struct {
+		TokenURL     string                 `mapstructure:"token_url"`
+		ClientID     string                 `mapstructure:"client_id"`
+		ClientSecret string                 `mapstructure:"client_secret"`
+		Username     string                 `mapstructure:"username"`
+		Password     string                 `mapstructure:"password"`
+		Rest         map[string]interface{} `mapstructure:",remain"`
+	}{}
+
+	if err := mapstructure.Decode(cfg, &decoded); err != nil {
+		return err
+	}
+
+	o.ClientID = decoded.ClientID
+	o.ClientSecret = decoded.ClientSecret
+	o.TokenURL = decoded.TokenURL
+	o.Username = decoded.Username
+	o.Password = decoded.Password
+
+	if err := o.validate(); err != nil {
+		return err
+	}
+
+	if len(decoded.Rest) > 0 {
+		return errors.New("unexpected fields in config")
```

Comment:
Why should we hard fail here if the Rest bytes are supplied.

We should just Warn and continue if the required essential parameters are valid and satisfy the constraints which we set.

Reply:
This is library code, not application code -- we should let the calling code handle it (it may have ideosyncratic mechansims for logging/warning).

Please note that the input here does not come directly from the end-user, it is a map generated by the calling code, so it's on that code to make sure the map is correct. It is up to the calling code to decide whether to allow additional fields in the input (and just ignore them when populating the map, or remove them if the map is generated from end-user input), or to report the error and fail. We should not be making that decision in the library code.